cli/flox-activations/src/cli/start_or_attach.rs

```diff
@@ -22,113 +25,404 @@ pub struct StartOrAttachArgs {
+        //
+        // TODO: just call StartOrAttach::handle again
+        anyhow::bail!(indoc! {"
+            Prior activation of the environment failed to start, or completed.
+
+            Try again to start a new activation of the environment.
+        "});
+    }
+
+    if now > attach_expiration {
+        anyhow::bail!(indoc! {"
+            Timed out waiting for a prior activation of the environment
+            to complete startup hooks.
+
+            Try again after the previous activation of the environment has completed.
+        "});
+    }
+    Ok(false)
+}
+
+#[cfg(test)]
+mod tests {
+    use uuid::Uuid;
+
+    use super::*;
+    use crate::cli::test::{read_activations, write_activations};
+
+    #[test]
+    fn attach_if_activation_exists() {
+        let cache_dir = tempfile::tempdir().unwrap();
+        let flox_env = PathBuf::from("/path/to/floxenv");
+        let store_path = "/store/path";
+
+        // The PID of the current process, guaranteed to be running
+        let pid = std::process::id();
+
+        let id = write_activations(&cache_dir, &flox_env, |activations| {
+            activations.create_activation(store_path, pid).unwrap().id()
+        });
+
+        let args = StartOrAttachArgs {
+            pid,
+            flox_env: flox_env.clone(),
+            store_path: store_path.to_string(),
         };
 
-        let activation = activations
-            .activation_for_id_mut(activation_id)
-            .context(indoc! {"
-                Prior activation of the environment completed.
+        let mut output = Vec::new();
 
-                Try again to start a new activation of the environment.
-            "})?;
+        args.handle_inner(
+            cache_dir.path().to_path_buf(),
+            |_, _, _, _| Ok(()),
+            |_, _, _, _, _| panic!("start should not be called"),
+            &mut output,
+        )
+        .expect("handle_inner should succeed");
 
-        if activation.ready() {
-            activation.attach_pid(attaching_pid, None);
-            activations::write_activations_json(&activations, activations_json_path, lock)?;
-            break;
-        }
+        let output = String::from_utf8(output).unwrap();
 
-        if !activation.startup_process_running() {
-            // TODO: clean out old activation of store_path
-            // Or we may need to do that in activation_for_store_path()
-            //
-            // TODO: just call StartOrAttach::handle again
-            anyhow::bail!(indoc! {"
-                Prior activation of the environment failed to start, or completed.
+        assert!(output.contains("_FLOX_ATTACH=true"));
+        assert!(output.contains(&format!(
+            "_FLOX_ACTIVATION_STATE_DIR={}",
+            activations::activation_state_dir_path(&cache_dir, flox_env, id)
+                .unwrap()
+                .display()
+        )));
+        assert!(output.contains(&format!("_FLOX_ACTIVATION_ID={}", id)));
```

Comment:
nonblocking: are you thinking injecting stdout is a better pattern than using `History` for crates other than `flox` and `flox-rust-sdk`?